executor/analyze.go

```diff
@@ -641,6 +641,56 @@ func (e *AnalyzeFastExec) getSampRegionsRowCount(bo *tikv.Backoffer, needRebuild
 	}
 }
 
+// getNextSampleKey gets the next sample key after last failed request. It only retries the needed region.
+// Different from other requests, each request range must be the whole region because the region row count
+// is only for a whole region. So we need to first find the longest successive prefix ranges of previous request,
+// then the next sample key should be the last range that could align with the region bound.
+func (e *AnalyzeFastExec) getNextSampleKey(bo *tikv.Backoffer, startKey kv.Key) (kv.Key, error) {
+	if len(e.sampTasks) == 0 {
+		e.scanTasks = e.scanTasks[:0]
+		return startKey, nil
+	}
+	sort.Slice(e.sampTasks, func(i, j int) bool {
+		return bytes.Compare(e.sampTasks[i].Location.StartKey, e.sampTasks[j].Location.StartKey) < 0
+	})
+	// The sample task should be consecutive with scan task.
+	if len(e.scanTasks) > 0 && bytes.Equal(e.scanTasks[0].StartKey, startKey) {
+		if !bytes.Equal(e.scanTasks[0].EndKey, e.sampTasks[0].Location.StartKey) {
+			e.scanTasks = e.scanTasks[:0]
+			e.sampTasks = e.sampTasks[:0]
+			return startKey, nil
+		}
+	}
+	prefixLen := 0
+	for ; prefixLen < len(e.sampTasks)-1; prefixLen++ {
+		if !bytes.Equal(e.sampTasks[prefixLen].Location.EndKey, e.sampTasks[prefixLen+1].Location.StartKey) {
```

Comment:
If the ranges of the keys in table are like `[1,3], [7,9]`, and these 2 ranges are located in 2 regions, would the region boundaries of the 2 regions be like `[1,5), [5,10)`? or the region boundaries would match the real ranges so there can be holes in the region boundaries?

Reply:
No, there will be no holes in the sample range and region range.